lib_eio/unix/thread_pool.ml

```diff
@@ -0,0 +1,155 @@
+  if Array.unsafe_get pool.available 5 = b then 5 else
+  if Array.unsafe_get pool.available 6 = b then 6 else
+  if Array.unsafe_get pool.available 7 = b then 7 else
+  if Array.unsafe_get pool.available 8 = b then 8 else
+  if Array.unsafe_get pool.available 9 = b then 9 else
+  if Array.unsafe_get pool.available 10 = b then 10 else
+  if Array.unsafe_get pool.available 11 = b then 11 else
+  if Array.unsafe_get pool.available 12 = b then 12 else
+  if Array.unsafe_get pool.available 13 = b then 13 else
+  if Array.unsafe_get pool.available 14 = b then 14 else
+  if Array.unsafe_get pool.available 15 = b then 15 else
+  if Array.unsafe_get pool.available 16 = b then 16 else
+  if Array.unsafe_get pool.available 17 = b then 17 else
+  if Array.unsafe_get pool.available 18 = b then 18 else
+  if Array.unsafe_get pool.available 19 = b then 19 else
+  raise Assertion_failure (* Can't [assert false] because it would allocate *)
+
+(* [@poll error] makes this function atomic from the point of view
+   of threads within a single domain.
+   This function either terminates the current systhread,
+   or adds/readds it to the list of available systhreads. *)
+let[@poll error] keep_thread_or_exit pool mbox =
+  if pool.terminating || pool.num_available = max_standby_systhreads_per_domain
+  then raise Thread.Exit
+  else (
+    let i = find_thread_predicate_exists pool false in
+    Array.unsafe_set pool.available i true;
+    Array.unsafe_set pool.threads i mbox;
+    pool.num_available <- pool.num_available + 1
+  )
+
+let make_thread pool =
+  let mbox = Mailbox.create () in
+  let _t : Thread.t = Thread.create (fun () ->
+    while true do
+      match Mailbox.take mbox with
+      | New -> assert false
+      | Exit -> raise Thread.Exit
+      | Job { fn; enqueue } ->
+        enqueue (try Ok (fn ()) with exn -> Error exn);
+        keep_thread_or_exit pool mbox
+    done
+  ) ()
+  in
+  mbox
+
+(* [@poll error] makes this function atomic from the point of view
+   of threads within a single domain.
+   This function tries to reserve one available thread from the pool.
+   Since we cannot return an option in [@poll error] code, we simulate
+   an option by conditionally updating the reference [res].
+   The returned boolean indicates whether we updated [res] or not. *)
+let[@poll error] try_get_thread pool res =
+  match pool.num_available with
+  | 0 -> false
+  | _ ->
+    let i = find_thread_predicate_exists pool true in
+    Array.unsafe_set pool.available i false;
+    res := Array.unsafe_get pool.threads i;
+    pool.num_available <- pool.num_available - 1;
+    true
+
+(* https://v2.ocaml.org/manual/parallelism.html#s:par_systhread_interaction
+   "Only one systhread at a time is allowed to run OCaml code on a particular domain."
+   So we keep a separate threadpool per domain. *)
+let key =
+  Domain.DLS.new_key @@ fun () ->
+    let pool = create () in
+    Domain.at_exit (fun () -> terminate pool);
```

Comment:
This is unfortunately problematic, because [`Domain.DLS` is currently not thread-safe](https://github.com/ocaml/ocaml/issues/12677) and, if I'd redesign it, I'd make it lock-free and specify that the init function (this function) passed to `Domain.DLS.new_key` may be called multiple times for a domain (but all concurrent accesses will eventually return only one of the results returned by the init function calls).  This means that anything that has side-effects that should be performed only once cannot be done inside the init function.

So, more specifically, it is possible that `Domain.at_exit` is called more than once and should not be done in the init function.  In [a Picos example](https://github.com/ocaml-multicore/picos/blob/7df26128a555edd9c69fba6199dc6eafd9e7b109/test/lib/elements/async_unix.ml#L131), I used a systhread safe locking approach (using `poll error` to perform initialization just once.

Currently `Domain.DLS` is completely unsafe with respect to systhreads (accesses of unrelated keys can mess things up), but I would hope that in the future it will be safe to allocate a new mutable object in the init function (currently it is not &mdash; nothing in `Domain.DLS` is currently thread-safe) and other systhread concurrency issues will be fixed.

Reply:
Thank you for this information. I'm definitely surprised by this, as the mli for `Domain.DLS` even shows an example that combines `Domain.DLS.new_key` with `Domain.at_exit`, but as we've discussed elsewhere both those features have deep issues that I need to work around.